File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.js
function solution(X, Y) {
    const map = new Map();
    
    for(let i = 0 ; i < X.length; i++){
        const x = X[i];
        
        const value = map.get(x);
        
        if(value){
            map.set(x, value + 1);
        } else {
            map.set(x, 1);
        }
    }
    
    const intersection = [];
    
    for(let i = 0; i < Y.length; i++){
        const y = Y[i];
        
        const value = map.get(y)
        if(value){
            intersection.push(y);
            map.set(y, value - 1);
        }
    }
    
    if(intersection.length === 0) return "-1";
    
    if(intersection.filter(i => i !== "0").length === 0) return "0";
    
    return intersection.sort((a, b) => b - a).join('');
    
}

File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.py
def solution(X, Y):
    x_dict = {}
    y_dict = {}
    
    for x in str(X):
        if x_dict.get(x):
            x_dict[x] += 1
        else:
            x_dict[x] = 1
    for y in str(Y):
        if y_dict.get(y):
            y_dict[y] += 1
        else:
            y_dict[y] = 1
            
    
    common = []
    
    [less_list, larger_list] = [x_dict, y_dict] if len(x_dict) < len(y_dict) else [y_dict, x_dict]
    
    for key in less_list.keys():
        if larger_list.get(key):
            for i in range(min(less_list[key], larger_list[key])):
                common.append(key)
    
    if len(common) == 0:
        return "-1"
    
    if any(c != '0' for c in common):
        return "".join(sorted(common, reverse=True))
    else:
        return "0"